validator_client/slashing_protection/src/slashing_protection.rs

```diff
@@ -0,0 +1,378 @@
+
+    fn verify_and_get_index(
+        &self,
+        incoming_data: &BeaconBlockHeader,
+        data_history: &[SignedBlock],
+    ) -> Result<Safe, NotSafe> {
+        check_for_proposer_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
+        let check = self.verify_and_get_index(incoming_data, &self.data[..]);
+        match check {
+            Ok(safe) => match safe.reason {
+                // Casting the same vote, no need to add it go the history
+                ValidityReason::SameVote => Ok(()),
+                // New attestation, add it to memory and file history
+                _ => self
+                    .insert_and_write(SignedBlock::from(incoming_data), safe.insert_index)
+                    .map_err(|e| NotSafe::IOError(e.kind())),
+            },
+            Err(notsafe) => Err(notsafe),
+        }
+    }
+}
+
+/// Struct used for checking if attestations or blockheader are safe from slashing.
+#[derive(Debug)]
+pub struct HistoryInfo<T: Encode + Decode + Clone + PartialEq> {
+    filepath: PathBuf,
+    data: Vec<T>,
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> PartialEq for HistoryInfo<T> {
+    fn eq(&self, other: &Self) -> bool {
+        let my_data = &self.data;
+        let other_data = &other.data;
+        self.filepath == other.filepath && *my_data == *other_data
+    }
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> HistoryInfo<T> {
+    /// Inserts the incoming data in the in-memory history, and writes it to the history file.
+    pub fn insert_and_write(&mut self, data: T, index: usize) -> IOResult<()> {
+        self.data.insert(index, data);
+
+        // Creating file if it doesn't exist, else opening it.
+        let mut file = File::create(self.filepath.as_path())?;
+
+        // Locking the file to make sure we're atomically writing to it
+        file.lock_exclusive()?;
+
+        // Setting permissions to be 600 (rw-------).
+        let mut perm = file.metadata()?.permissions();
+        perm.set_mode(0o600);
+        file.set_permissions(perm)?;
+
+        // Writing new history to file
+        file.write_all(&self.data.as_ssz_bytes())?;
+
+        // Unlocking file because we're done with it.
+        file.unlock()?;
+
+        Ok(())
+    }
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> TryFrom<PathBuf> for HistoryInfo<T> {
+    type Error = NotSafe;
+
+    fn try_from(filepath: PathBuf) -> Result<Self, Self::Error> {
+        let path = filepath.as_path();
+        Self::try_from(path)
+    }
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> TryFrom<&Path> for HistoryInfo<T> {
+    type Error = NotSafe;
+
+    fn try_from(filepath: &Path) -> Result<Self, Self::Error> {
+        let mut file = match File::open(filepath) {
+            Ok(file) => file,
+            Err(e) => match e.kind() {
+                // File was not found meaning it has not been created yet, and that history is empty.
```

Comment:
Are we sure this is the behaviour we want? Or would we rather create the file in advance, and returning an error here if file was not found (i.e assuming that the file already exists)